### PR TITLE
feat(watermark): clean state for top n by watermark

### PR DESCRIPTION
**Pull request by yuhao-su**
I hereby agree to the terms of the [RisingWave Labs, Inc. Contributor License Agreement](https://gist.github.com/TennyZhuang/f00be7f16996ea48effb049aa7be4d66#file-rw_cla).

## What's changed and what's your intention?

- clean state for top n by watermark

#7810 
#8106

TODO:
add tests.

## Checklist For Contributors

- [x] I have demonstrated that backward compatibility is not broken by breaking changes and created issues to track deprecated features to be removed in the future. (Please refer to the issue)
- [x] All checks passed in `./risedev check` (or alias, `./risedev c`)

## Checklist For Reviewers

- [ ] I have requested macro/micro-benchmarks as this PR can affect performance substantially, and the results are shown.

## Documentation

- [x] My PR **DOES NOT** contain user-facing changes.

<details><summary>Click here for Documentation</summary>

### Types of user-facing changes

Please keep the types that apply to your changes, and remove the others.

- Installation and deployment
- Connector (sources & sinks)
- SQL commands, functions, and operators
- RisingWave cluster configuration changes
- Other (please specify in the release note below)

### Release note

</details>

### CODE REVIEW

File: src/stream/src/executor/top_n/top_n_appendonly.rs [hunk 1]
```diff
@@ -188,9 +188,15 @@ where
...
 
-    async fn handle_watermark(&mut self, _: Watermark) -> Option<Watermark> {
-        // TODO(yuhao): handle watermark
-        None
+    async fn handle_watermark(
+        &mut self,
+        watermark: Watermark,
+    ) -> StreamExecutorResult<Option<Watermark>> {
+        Ok(if watermark.col_idx == self.storage_key_indices[0] {
+            Some(watermark)
```
Comment (st1page) [inline src/stream/src/executor/top_n/top_n_appendonly.rs:196]: ditto. descending case. I think we need to compare the last key with the watermark here.

**Review by st1page — COMMENTED** [review]
Please add the logic of watermark derivation in optimizer too.

File: src/stream/src/executor/top_n/top_n_plain.rs [hunk 2]
```diff
@@ -228,9 +228,34 @@ where
...
+                            // TODO(yuhao): range delete records above the watermark
+                        }
+                    } else {
+                        // TODO(yuhao): handle null watermark
+                    }
+                }
+            } else {
+                // TODO(yuhao): handle watermark in descending case
+            }
+            Ok(Some(watermark))
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/stream/src/executor/top_n/top_n_plain.rs:255]: Oh not like this. Will fix.

File: src/stream/src/executor/top_n/top_n_plain.rs [hunk 3]
```diff
@@ -228,9 +228,34 @@ where
...
+                        .cache_key_serde
+                        .0
+                        .deserialize_first(&last_entry.key().0)?;
+
+                    if let Some(watermark_in_key) = watermark_in_key {
+                        if watermark.val > watermark_in_key {
+                            // TODO(yuhao): range delete records above the watermark
+                        }
+                    } else {
+                        // TODO(yuhao): handle null watermark
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/stream/src/executor/top_n/top_n_plain.rs:249]: Yes. I should name it event time column

File: src/stream/src/executor/top_n/top_n_plain.rs [hunk 4]
```diff
@@ -228,9 +228,34 @@ where
...
+                if self.cache.is_middle_cache_full() {
+                    let last_entry = self.cache.middle.last_entry().unwrap();
+                    let watermark_in_key = self
+                        .cache_key_serde
+                        .0
+                        .deserialize_first(&last_entry.key().0)?;
+
+                    if let Some(watermark_in_key) = watermark_in_key {
+                        if watermark.val > watermark_in_key {
+                            // TODO(yuhao): range delete records above the watermark
```
[the remark this answers is not in the document]
Comment (yuhao-su) [reply src/stream/src/executor/top_n/top_n_plain.rs:246]: Yes. It should work like the current `update_watermark` but remove values below watermark instead above.

**Pull request closed without merging**